File: src/main/java/com/indiepost/service/StaticPageServiceImpl.java
package com.indiepost.service;

import com.indiepost.dto.StaticPageDto;
import com.indiepost.enums.Types;
import com.indiepost.model.StaticPage;
import com.indiepost.model.User;
import com.indiepost.repository.StaticPageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Caching;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

import static com.indiepost.utils.DateUtil.localDateTimeToInstant;

/**
 * Created by jake on 17. 3. 5.
 */
@Service
@Transactional
public class StaticPageServiceImpl implements StaticPageService {

    private final StaticPageRepository staticPageRepository;

    private final UserService userService;

    @Autowired
    public StaticPageServiceImpl(StaticPageRepository staticPageRepository, UserService userService) {
        this.staticPageRepository = staticPageRepository;
        this.userService = userService;
    }

    @Override
    @CacheEvict(cacheNames = "static-page::rendered", key = "#dto.slug")
    public Long save(StaticPageDto dto) {
        StaticPage staticPage = new StaticPage();
        staticPage.setTitle(dto.getTitle());
        staticPage.setContent(dto.getContent());
        staticPage.setSlug(dto.getSlug());
        staticPage.setDisplayOrder(dto.getDisplayOrder());
        staticPage.setType(dto.getType());
        staticPage.setStatus(dto.getStatus());

        User currentUser = userService.findCurrentUser();
        staticPage.setAuthor(currentUser);

        staticPage.setCreatedAt(LocalDateTime.now());
        staticPage.setModifiedAt(LocalDateTime.now());
        return staticPageRepository.save(staticPage);
    }

    @Override
    @Caching(evict = {
            @CacheEvict(cacheNames = "static-page::rendered", key = "#dto.slug"),
            @CacheEvict(cacheNames = "home::rendered", allEntries = true)
    })
    public void update(StaticPageDto dto) {
        StaticPage staticPage = staticPageRepository.findById(dto.getId());
        staticPage.setTitle(dto.getTitle());
        staticPage.setContent(dto.getContent());
        staticPage.setSlug(dto.getSlug());
        staticPage.setDisplayOrder(dto.getDisplayOrder());
        staticPage.setType(dto.getType());
        staticPage.setStatus(dto.getStatus());

        staticPage.setModifiedAt(LocalDateTime.now());
        staticPageRepository.update(staticPage);
    }

    @Override
    public StaticPageDto findById(Long id) {
        StaticPage staticPage = staticPageRepository.findById(id);
        return pageToPageDto(staticPage);
    }

    @Override
    public StaticPageDto findBySlug(String slug) {
        StaticPage staticPage = staticPageRepository.findBySlug(slug);
        return pageToPageDto(staticPage);
    }

    @Override
    public Page<StaticPageDto> find(Pageable pageable) {
        return staticPageRepository.find(addOrder(pageable));
    }

    @Override
    public Page<StaticPageDto> find(Types.PostStatus pageStatus, Pageable pageable) {
        return staticPageRepository.find(addOrder(pageable), pageStatus);
    }

    @Override
    public void delete(StaticPageDto staticPageDto) {
        StaticPage staticPage = staticPageRepository.findById(staticPageDto.getId());
        staticPageRepository.delete(staticPage);
    }

    @Override
    public void deleteById(Long id) {
        StaticPage staticPage = staticPageRepository.findById(id);
        staticPageRepository.delete(staticPage);
    }

    @Override
    public Long count() {
        return staticPageRepository.count();
    }

    @Override
    public void bulkUpdateStatus(List<Long> ids, Types.PostStatus status) {
        staticPageRepository.bulkUpdateStatusByIds(ids, status);
    }

    @Override
    public void bulkDeleteByIds(List<Long> ids) {
        staticPageRepository.bulkDeleteByIds(ids);
    }

    @Override
    public void emptyTrash() {
        staticPageRepository.bulkDeleteByStatus(Types.PostStatus.TRASH);
    }

    private StaticPageDto pageToPageDto(StaticPage staticPage) {
        StaticPageDto staticPageDto = new StaticPageDto();
        staticPageDto.setId(staticPage.getId());
        staticPageDto.setTitle(staticPage.getTitle());
        staticPageDto.setContent(staticPage.getContent());
        staticPageDto.setType(staticPage.getType());
        staticPageDto.setAuthorDisplayName(staticPage.getAuthor().getDisplayName());
        staticPageDto.setSlug(staticPage.getSlug());
        staticPageDto.setDisplayOrder(staticPage.getDisplayOrder());
        staticPageDto.setModifiedAt(localDateTimeToInstant(staticPage.getModifiedAt()));
        staticPageDto.setCreatedAt(localDateTimeToInstant(staticPage.getCreatedAt()));
        staticPageDto.setStatus(staticPage.getStatus());
        return staticPageDto;
    }

    private Pageable addOrder(Pageable pageable) {
        return PageRequest.of(pageable.getPageNumber(), pageable.getPageSize(), Sort.Direction.ASC, "displayOrder");
    }
}

File: src/main/kotlin/com/indiepost/service/StaticPageServiceImpl.kt
package com.indiepost.service

import com.indiepost.dto.StaticPageDto
import com.indiepost.enums.Types
import com.indiepost.exceptions.ResourceNotFoundException
import com.indiepost.model.StaticPage
import com.indiepost.repository.StaticPageRepository
import com.indiepost.utils.DateUtil.localDateTimeToInstant
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Caching
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import javax.inject.Inject

/**
 * Created by jake on 17. 3. 5.
 */
@Service
@Transactional
class StaticPageServiceImpl @Inject
constructor(private val staticPageRepository: StaticPageRepository, private val userService: UserService) : StaticPageService {

    @CacheEvict(cacheNames = ["static-page::rendered"], key = "#dto.slug")
    override fun save(dto: StaticPageDto): Long {
        val staticPage = StaticPage()
        staticPage.title = dto.title
        staticPage.content = dto.content
        staticPage.slug = dto.slug
        staticPage.displayOrder = dto.displayOrder
        staticPage.type = dto.type
        staticPage.status = dto.status ?: Types.PostStatus.DRAFT

        val currentUser = userService.findCurrentUser()
        staticPage.author = currentUser

        staticPage.createdAt = LocalDateTime.now()
        staticPage.modifiedAt = LocalDateTime.now()
        return staticPageRepository.save(staticPage)
    }

    @Caching(evict = [
        CacheEvict(cacheNames = arrayOf("static-page::rendered"), key = "#dto.slug"),
        CacheEvict(cacheNames = arrayOf("home::rendered"), allEntries = true)])
    override fun update(dto: StaticPageDto) {
        val staticPage = staticPageRepository.findById(dto.id!!)
                ?: throw ResourceNotFoundException("No staticPage with id: " + dto.id)
        staticPage.title = dto.title
        staticPage.content = dto.content
        staticPage.slug = dto.slug
        staticPage.displayOrder = dto.displayOrder
        staticPage.type = dto.type
        dto.status?.let {
            staticPage.status = it
        }
        staticPage.modifiedAt = LocalDateTime.now()
        staticPageRepository.update(staticPage)
    }

    override fun findById(id: Long): StaticPageDto? {
        val staticPage = staticPageRepository.findById(id)
        return if (staticPage != null)
            pageToPageDto(staticPage)
        else null
    }

    override fun findBySlug(slug: String): StaticPageDto {
        val staticPage = staticPageRepository.findBySlug(slug) ?: throw ResourceNotFoundException()
        return pageToPageDto(staticPage)
    }

    override fun find(pageable: Pageable): Page<StaticPageDto> {
        return staticPageRepository.find(addOrder(pageable))
    }

    override fun find(status: Types.PostStatus, pageable: Pageable): Page<StaticPageDto> {
        return staticPageRepository.find(addOrder(pageable), status)
    }

    override fun delete(staticPageDto: StaticPageDto) {
        val id = staticPageDto.id ?: return
        val staticPage = staticPageRepository.findById(id) ?: return
        staticPageRepository.delete(staticPage)
    }

    override fun deleteById(id: Long) {
        val staticPage = staticPageRepository.findById(id) ?: return
        staticPageRepository.delete(staticPage)
    }

    override fun count(): Long {
        return staticPageRepository.count()
    }

    override fun bulkUpdateStatus(ids: List<Long>, status: Types.PostStatus) {
        staticPageRepository.bulkUpdateStatusByIds(ids, status)
    }

    override fun bulkDeleteByIds(ids: List<Long>) {
        staticPageRepository.bulkDeleteByIds(ids)
    }

    override fun emptyTrash() {
        staticPageRepository.bulkDeleteByStatus(Types.PostStatus.TRASH)
    }

    private fun pageToPageDto(staticPage: StaticPage): StaticPageDto {
        val staticPageDto = StaticPageDto()
        staticPageDto.id = staticPage.id
        staticPageDto.title = staticPage.title
        staticPageDto.content = staticPage.content
        staticPageDto.type = staticPage.type
        staticPageDto.authorDisplayName = staticPage.author!!.displayName
        staticPageDto.slug = staticPage.slug
        staticPageDto.displayOrder = staticPage.displayOrder
        staticPageDto.modifiedAt = localDateTimeToInstant(staticPage.modifiedAt!!)
        staticPageDto.createdAt = localDateTimeToInstant(staticPage.createdAt!!)
        staticPageDto.status = staticPage.status
        return staticPageDto
    }

    private fun addOrder(pageable: Pageable): Pageable {
        return PageRequest.of(pageable.pageNumber, pageable.pageSize, Sort.Direction.ASC, "displayOrder")
    }
}
